### PR TITLE
Add spec for monitoring architecture

### PR DESCRIPTION
**Pull request by anmolbabu**
tendrl-bug-id: Tendrl/specifications#71
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

### CODE REVIEW

**Comment by anmolbabu** [comment]
@shtripat @nthomas-redhat @brainfunked @r0h4n please review

**Comment by anmolbabu** [comment]
I tried using the ascii flow for the diagram but then the generated diagram is loaded as normal text(not image) and hence when it is loaded in adoc it violates the 79 char limit and looks odd in asciidoctor generated  document. So using the png itself in PR

**Comment by r0h4n** [comment]
@anmolbabu  for review requests use below method

@Tendrl/tendrl-core @Tendrl/tendrl_api @Tendrl/qe 
Please review

**Review by brainfunked — APPROVED** [review]

**Review by cybernth — APPROVED** [review]

File: specs/monitoring_architecture.adoc [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+
+* Performance monitoring
+* Status monitoring
+
+The status monitoring can be further classified as:
+
+* Process status monitoring
+* Resource status monitoring
+
+The essential functionalities provided by any performance moniotoring systems
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:19]: %s/moniotoring/monitoring/g

File: specs/monitoring_architecture.adoc [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+are:
+
+* Statistics collection
+* Thresholding on the collected statistics
+* Alerting/notifying the users about the threshold breach if any.
+
+The essential functionalities provided by any status moniotoring systems are:
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:26]: %s/moniotoring/monitoring/g

File: specs/monitoring_architecture.adoc [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+* Event listeners to listen to change in status of the resource if provided by
+  the sds framework or a continous/scheduled hearbeat to monitor changes in
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:29]: %s/continous/continuous/g

File: specs/monitoring_architecture.adoc [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+image::tendrl_monitoring_components_overview.png[Tendrl Monitoring Components]
+
+=== Overview Of Components Of Monitoring Stack
+
+==== Tendrl Performance (performance_monitoring)
+
+It is a separate service that runs outside tendrl core stack and it relies on
+the state information in tendrl’s central store for its functioning. Typically
+it can be co-resident with tendrl application but nothing necessiates this.
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:71]: %s/necessiates/necessitates/g

File: specs/monitoring_architecture.adoc [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+  Note:
+  * Detailed description can be found in:
+      https://github.com/Tendrl/specifications/pull/18
+--------
+
+==== Tendrl Alerts (tendrl_alerts)
+
+* A separate service that is installed with Tendrl App.
+* It serves the following responsibilities:
+  ** Watch for new alert/notification details in central store.
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:93]: No indentation required here

File: specs/monitoring_architecture.adoc [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+    *** This application utilizes the etcd's watch capability to sense changes
+        in etcd.
+  ** Send out alerts to users configured to receive them. Following are the
+     currently supported means of notifying alerts to configured users:
+    *** Email alerting
+    *** Snmp alerting
+  ** The delivery gurantees will be in accordance with the significance of the
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:100]: %s/gurantees/guarantees/g

File: specs/monitoring_architecture.adoc [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+  with the following:
+  ** Monitoring application triggers the monitoring provisioning flow on every
+     node for node monitoring via the core API.
+  ** The provisioning flow for the monitoring stack installs collectd-related
+     pieces on every node and generates the configuration files from templates
+     shipped with the collectd plugins on nodes.
+* The configured plugins start collecting the respective utilizations at every
+  configured intervals of time and push them to the write destination using
+  the plugin configured as write plugin.
+* Parallely, every configured intervals of time the aggregator module of the
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:191]: %s/Parallely/Parallelly/g

File: specs/monitoring_architecture.adoc [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+  central monitoring application will collect the instant value of stats from
+  the time series db and update the stats of interest to tendrl into tendrl’s
+  central store and also aggregates these stats @ cluster and system levels
```
Comment (shtripat) [inline specs/monitoring_architecture.adoc:194]: Avoid using slangs :)

**Review by shtripat — COMMENTED** [review]
Looks good to me but some spelling mistakes identified...

**Pull request merged**